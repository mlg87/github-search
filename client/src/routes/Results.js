import React, { Component } from 'react'
import { inject, observer } from 'mobx-react'

import '../styles/Results.css'

const Results = inject('repositoryStore')(
  observer(
    class Results extends Component {
      renderResults() {
        return this.props.repositoryStore.results.map(
          (
            { name, description, stargazers_count, language, owner: { login } },
            i
          ) => {
            return (
              <li
                key={`${name}${i}`}
              >{`${name} | ${description} | ${stargazers_count} | ${language} | ${login}`}</li>
            )
          }
        )
      }

      render() {
        const { results } = this.props.repositoryStore

        return (
          <div className="results-container">
            {!results.length ? (
              <p>No results to display</p>
            ) : (
              <ul>{this.renderResults()}</ul>
            )}
          </div>
        )
      }
    }
  )
)

export default Results
