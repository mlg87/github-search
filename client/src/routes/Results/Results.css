.results {
  display: flex;
  flex-grow: 1;
  justify-content: center;
  align-items: center;
}

.results__message {
  color: rgb(194, 194, 194);
}

.results__table {
  width: 100%;
  table-layout: fixed;
  border-collapse: collapse;
}

.results__th {
  padding: 20px 15px;
  text-align: left;
  font-weight: 500;
  font-size: 12px;
  color: #000;
  text-transform: uppercase;
  cursor: pointer;
}

.results__thead-border {
  height: 1px;
  width: 100vw;
  background: linear-gradient(to right, #f09a38, #d888e2);
}

.results__sort-order {
  margin-left: 10px;
}

.results__td {
  padding: 15px;
  text-align: left;
  vertical-align: middle;
  font-weight: 300;
  font-size: 12px;
  color: #000;
  border-bottom: solid 1px rgb(194, 194, 194);
}

.results__content {
  /* 160px to give a little padding */
  height: calc(100vh - 160px);
  overflow-x: auto;
  margin-top: 0px;
}
