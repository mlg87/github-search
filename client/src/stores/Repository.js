import { observable, action, decorate } from 'mobx'
import API from '../utils/api'

class RepositoryStore extends API {
  results = []
  query = ''
  isFetching = false
  error = null

  async fetchSearch() {
    this.isFetching = true
    try {
      const q = this.query.split(' ').reduce((accum, cv, i) => {
        return i === 0 ? cv : `${accum}+${cv}`
      }, '')

      const res = await this.fetch(`repositories?q=${q}`)
      const { error, content } = await res.json()

      if (res.status !== 200) throw Error(error)

      this.results = content.items

      return null
    } catch (error) {
      this.error = error
    } finally {
      this.isFetching = false
    }
  }
}

decorate(RepositoryStore, {
  error: observable,
  isFetching: observable,
  query: observable,
  results: observable
})

export default new RepositoryStore()
