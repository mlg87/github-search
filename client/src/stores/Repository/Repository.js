import { observable, action, decorate } from 'mobx'
import orderBy from 'lodash/orderBy'
import API from '../../utils/api'

class RepositoryStore extends API {
  //
  // ─── OBSERVABLES ────────────────────────────────────────────────────────────────
  //
  results = []
  query = ''
  sortByStars = false
  resultsSortBy = 'relevance'
  resultsSortOrder = 'asc'
  isFetching = false
  error = null

  //
  // ─── ACTIONS ────────────────────────────────────────────────────────────────────
  //
  async fetchSearch() {
    this.isFetching = true
    this.error = null
    try {
      const q = this.query.split(' ').reduce((accum, cv, i) => {
        return i === 0 ? cv : `${accum}+${cv}`
      }, '')

      const sort = this.sortByStars ? 'stars' : 'default'

      const res = await this.fetch(`repositories?q=${q}&sort=${sort}`)
      const { error, content } = await res.json()

      if (res.status !== 200) throw Error(error)

      this.results = content.items

      // if user wants results sorted by stars, set table sort observables
      // otherwise, let it be sorted by relevance (not currently a col)
      if (this.sortByStars) {
        this.resultsSortBy = 'stargazers_count'
        this.resultsSortOrder = 'desc'
      }

      return null
    } catch (error) {
      this.error = error
      throw error
    } finally {
      this.isFetching = false
    }
  }

  toggleSort(isChecked) {
    return (this.sortByStars = isChecked)
  }

  sortResults(col) {
    // reset order if new col
    if (this.resultsSortBy !== col) {
      this.resultsSortOrder = 'asc'
    } else {
      this.resultsSortOrder = this.resultsSortOrder === 'asc' ? 'desc' : 'asc'
    }
    // set col we are sorting by
    this.resultsSortBy = col
    this.results = orderBy(
      this.results,
      this.resultsSortBy,
      this.resultsSortOrder
    )
  }
}

decorate(RepositoryStore, {
  error: observable,
  fetchSearch: action,
  isFetching: observable,
  query: observable,
  results: observable,
  resultsSortBy: observable,
  sortByStars: observable,
  sortResults: action,
  toggleSort: action
})

export default new RepositoryStore()
// for tests
export { RepositoryStore }
