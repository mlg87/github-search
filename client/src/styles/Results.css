.results {
  display: flex;
  flex-grow: 1;
  justify-content: center;
  align-items: center;
}

.results__header {
  background-color: rgba(255, 255, 255, 0.3);
}

.results__table {
  width: 100%;
  table-layout: fixed;
}

.results__th {
  padding: 20px 15px;
  text-align: left;
  font-weight: 500;
  font-size: 12px;
  color: #000;
  text-transform: uppercase;
}

.results__td {
  padding: 15px;
  text-align: left;
  vertical-align: middle;
  font-weight: 300;
  font-size: 12px;
  color: #000;
  border-bottom: solid 1px rgba(255, 255, 255, 0.1);
}

.results__content {
  /* 160px to give a little padding */
  height: calc(100vh - 160px);
  overflow-x: auto;
  margin-top: 0px;
  border: 1px solid rgba(255, 255, 255, 0.3);
}
