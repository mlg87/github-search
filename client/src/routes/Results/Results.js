import React, { Component } from 'react'
import { inject, observer } from 'mobx-react'
import SortedDown from 'react-icons/lib/ti/arrow-sorted-down'
import SortedUp from 'react-icons/lib/ti/arrow-sorted-up'

import './Results.css'

// TODO refactor from table to css grid to get rid of warnging about th border div
class Results extends Component {
  _handleColumnSort = col => {
    this.props.repositoryStore.sortResults(col)
  }

  renderResults() {
    return this.props.repositoryStore.results.map(
      (
        { name, description, stargazers_count, language, owner: { login } },
        i
      ) => {
        return (
          <tr key={`${name}-${i}`} className="results__tr">
            <td className="results__td">{name}</td>
            <td className="results__td">{description || 'n/a'}</td>
            <td className="results__td">{login}</td>
            <td className="results__td">{language || 'n/a'}</td>
            <td className="results__td">{stargazers_count}</td>
          </tr>
        )
      }
    )
  }

  renderTableHeader() {
    const { resultsSortBy, resultsSortOrder } = this.props.repositoryStore
    const columns = [
      { col: 'name', display: 'name' },
      { col: 'description', display: 'description' },
      { col: 'owner.login', display: 'owner' },
      { col: 'language', display: 'language' },
      { col: 'stargazers_count', display: 'stars' }
    ]

    const columnHeaders = columns.map(({ col, display }) => {
      return (
        <th
          key={col}
          className="results__th"
          onClick={() => this._handleColumnSort(col)}
        >
          {display}
          {resultsSortBy === col && (
            <span className="results__sort-order">
              {resultsSortOrder === 'asc' ? <SortedUp /> : <SortedDown />}
            </span>
          )}
        </th>
      )
    })

    return (
      <div>
        <table
          className="results__table"
          cellPadding="0"
          cellSpacing="0"
          border="0"
        >
          <thead>
            <tr>{columnHeaders}</tr>
            <div className="results__thead-border" />
          </thead>
        </table>
      </div>
    )
  }

  renderTableContent() {
    return (
      <div className="results__content">
        <table className="results__table">
          <tbody>{this.renderResults()}</tbody>
        </table>
      </div>
    )
  }

  renderTable() {
    return (
      <div>
        {this.renderTableHeader()}
        {this.renderTableContent()}
      </div>
    )
  }

  render() {
    const { results } = this.props.repositoryStore

    return (
      <div className="results">
        {!results.length ? (
          <p className="results__message">No results to display</p>
        ) : (
          this.renderTable()
        )}
      </div>
    )
  }
}

export default inject('repositoryStore')(observer(Results))
// for tests
export { Results }
