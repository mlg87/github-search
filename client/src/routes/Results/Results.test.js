import React from 'react'
import { Results } from './Results'
import { RepositoryStore } from '../../stores/Repository/Repository'

describe('Results', () => {
  let store

  beforeEach(() => {
    store = new RepositoryStore()
  })

  test('renders correctly', () => {
    const wrapper = shallow(<Results repositoryStore={store} />)
    expect(wrapper).toMatchSnapshot()
  })
})
