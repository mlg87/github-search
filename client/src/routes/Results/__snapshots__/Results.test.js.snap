// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Results
    repositoryStore={
        RepositoryStore {
            "error": null,
            "isFetching": false,
            "query": "",
            "results": ObservableArray [],
            "resultsSortBy": "relevance",
            "resultsSortOrder": "asc",
            "sortByStars": false,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p
        className="results__message"
>
        No results to display
</p>,
      "className": "results",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "No results to display",
        "className": "results__message",
      },
      "ref": null,
      "rendered": "No results to display",
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p
          className="results__message"
>
          No results to display
</p>,
        "className": "results",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "No results to display",
          "className": "results__message",
        },
        "ref": null,
        "rendered": "No results to display",
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
